libp2p/module.go

```diff
@@ -0,0 +1,367 @@
+    - Network interface can be simplified
+    - Consider renaming network as it functions more like a "router"
+    (NB: could be replaced in future iterations with a "raintree pubsub router")
+ 3. Remove "legacy" P2P module & rename libp2p module directory (possibly object names as well)
+    - P2PModule interface can be simplified
+    - Clean up TECHDEBT introduced in debug CLI and node startup
+*/
+package libp2p
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/libp2p/go-libp2p"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/libp2p/go-libp2p/core/host"
+	libp2pNetwork "github.com/libp2p/go-libp2p/core/network"
+	"github.com/multiformats/go-multiaddr"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/pokt-network/pocket/libp2p/network"
+	"github.com/pokt-network/pocket/libp2p/protocol"
+	"github.com/pokt-network/pocket/logger"
+	typesP2P "github.com/pokt-network/pocket/p2p/types"
+	"github.com/pokt-network/pocket/runtime/configs"
+	"github.com/pokt-network/pocket/runtime/configs/types"
+	"github.com/pokt-network/pocket/shared/crypto"
+	"github.com/pokt-network/pocket/shared/messaging"
+	"github.com/pokt-network/pocket/shared/modules"
+	"github.com/pokt-network/pocket/shared/modules/base_modules"
+)
+
+var _ modules.P2PModule = &libp2pModule{}
+
+type libp2pModule struct {
+	base_modules.IntegratableModule
+
+	logger *modules.Logger
+	//nolint:unused // bus is used by embedded base module(s)
+	bus         modules.Bus
+	cfg         *configs.P2PConfig
+	identity    libp2p.Option
+	listenAddrs libp2p.Option
+	// host encapsulates libp2p peerstore & connection manager
+	host host.Host
+	// pubsub is used for broadcast communication
+	// (i.e. multiple, unidentified receivers)
+	pubsub *pubsub.PubSub
+	// topic similar to pubsub but received messages are filtered by a "topic" string.
+	// Published messages are also given the respective topic before broadcast.
+	topic *pubsub.Topic
+	// subscription provides an interface to continuously read messages from.
+	subscription *pubsub.Subscription
+	network      typesP2P.Network
+}
+
+var (
+	// TECHDEBT: configure timeouts. Consider security exposure vs. real-world conditions).
+	// TECHDEBT: parameterize and expose via config.
+	// readStreamTimeout is the duration to wait for a read operation on a
+	// stream to complete, after which the stream is closed ("timed out").
+	readStreamTimeoutDuration = time.Second * 10
+)
+
+func Create(bus modules.Bus, options ...modules.ModuleOption) (modules.Module, error) {
+	return new(libp2pModule).Create(bus, options...)
+}
+
+func (mod *libp2pModule) GetModuleName() string {
+	return modules.P2PModuleName
+}
+
+func (mod *libp2pModule) Create(bus modules.Bus, options ...modules.ModuleOption) (modules.Module, error) {
+	logger.Global.Print("Creating libp2p-backed network module")
+
+	*mod = libp2pModule{
+		cfg: bus.GetRuntimeMgr().GetConfig().P2P,
+	}
```

Comment:
How about:

```suggestion
func (*libp2pModule) Create(bus modules.Bus, options ...modules.ModuleOption) (modules.Module, error) {
	logger.Global.Debug().Msg("Creating libp2p-backed network module")
	mod := &libp2pModule{
		cfg:    bus.GetRuntimeMgr().GetConfig().P2P,
		logger: logger.Global.CreateLoggerForModule(modules.P2PModuleName),
	}
```
?

More consistent with other modules.

Reply:
:100: agreed. We need to update [the shared modules docs](https://github.com/pokt-network/pocket/tree/main/shared/modules/doc#add-a-logger-to-the-module) to reflect this new convention, as currently, it effectively says that logger initialization should happen in `#Start()`:

> When defining the start function for the module, it is essential to initialise a namespace logger as well:

Do you think this is covered by #509, specifically the first goal?:

> Add godoc comments to the various interfaces, functions, methods involved in the module creation process

Reply:
> 💯 agreed. We need to update [the shared modules docs](https://github.com/pokt-network/pocket/tree/main/shared/modules/doc#add-a-logger-to-the-module) to reflect this new convention, as currently, it effectively says that logger initialization should happen in `#Start()`:
> 
> > When defining the start function for the module, it is essential to initialise a namespace logger as well:

Totally, thanks for pointing that out!

> Do you think this is covered by #509, specifically the first goal?:
> 
> > Add godoc comments to the various interfaces, functions, methods involved in the module creation process

Agreed, I have clarified the scope so I don't forget :) Thanks buddy!

![image](https://user-images.githubusercontent.com/29378614/222151789-db59b899-0107-45df-b7c5-72f65007294d.png)